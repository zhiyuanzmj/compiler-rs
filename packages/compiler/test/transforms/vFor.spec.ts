import { describe, expect, test } from 'vitest'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
  transformVFor,
  transformVOn,
  transformVText,
  type ForIRNode,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithVFor = makeCompile({
  nodeTransforms: [
    transformVFor,
    transformElement,
    transformText,
    transformChildren,
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
    text: transformVText,
  },
})

describe('compiler: v-for', () => {
  test('basic v-for', () => {
    const { code, ir, helpers } = compileWithVFor(
      `<div v-for={item in items} key={item.id} onClick={() => remove(item)}>{item}</div>`,
    )

    expect(code).toMatchSnapshot()

    expect(helpers).contains('createFor')
    expect(ir.templates).toEqual(['<div> </div>'])
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      id: 0,
      source: {
        content: 'items',
      },
      value: {
        content: 'item',
      },
      key: undefined,
      index: undefined,
      render: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }],
        },
      },
      keyProp: {
        content: 'item.id',
      },
    })
    expect(ir.block.returns).toEqual([0])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 0 }],
    })
    expect(ir.block.effect).toEqual([])
  })

  test('key only binding pattern', () => {
    expect(
      compileWithVFor(
        `
          <tr
            v-for={row in rows}
            key={row.id}
          >
            { row.id + row.id }
          </tr>
      `,
      ).code,
    ).matchSnapshot()
  })

  test('selector pattern', () => {
    expect(
      compileWithVFor(
        `
          <tr
            v-for={row in rows}
            key={row.id}
            v-text={selected === row.id ? 'danger' : ''}
          ></tr>
      `,
      ).code,
    ).matchSnapshot()

    expect(
      compileWithVFor(
        `
          <tr
            v-for={row in rows}
            key={row.id}
            class={selected === row.id ? 'danger' : ''}
          ></tr>
      `,
      ).code,
    ).matchSnapshot()

    // Should not be optimized because row.label is not from parent scope
    expect(
      compileWithVFor(
        `
          <tr
            v-for={row in rows}
            key={row.id}
            class={row.label === row.id ? 'danger' : ''}
          ></tr>
      `,
      ).code,
    ).matchSnapshot()

    expect(
      compileWithVFor(
        `
          <tr
            v-for={row in rows}
            key={row.id}
            class={{ danger: row.id === selected }}
          ></tr>
      `,
      ).code,
    ).matchSnapshot()
  })

  test('multi effect', () => {
    const { code } = compileWithVFor(
      `<div v-for={(item, index) in items} item={item} index={index} />`,
    )
    expect(code).matchSnapshot()
  })

  test('nested v-for', () => {
    const { code, ir } = compileWithVFor(
      `<div v-for={i in list}><span v-for={j in i}>{ j+i }</span></div>`,
    )
    expect(code).matchSnapshot()
    expect(code).contains(`_createFor(() => (list), (_for_item0) => {`)
    expect(code).contains(
      `_createFor(() => (_for_item0.value), (_for_item1) => {`,
    )
    expect(code).contains(`_for_item1.value+_for_item0.value`)
    expect(ir.templates).toEqual(['<span> </span>', '<div></div>'])
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      id: 0,
      source: { content: 'list' },
      value: { content: 'i' },
      render: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 1 }],
        },
      },
    })
    expect(
      (op as ForIRNode).render.dynamic.children[0].children[0].operation,
    ).toMatchObject({
      type: IRNodeTypes.FOR,
      id: 2,
      source: { content: 'i' },
      value: { content: 'j' },
      render: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }],
        },
      },
    })
  })

  test('object value, key and index', () => {
    const { code, ir } = compileWithVFor(
      '<span v-for={(value, key, index) in items} key={id}>{ id }{ value }{ index }</span>',
    )
    expect(code).matchSnapshot()
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'items',
      },
      value: {
        content: 'value',
      },
      key: {
        content: 'key',
      },
      index: {
        content: 'index',
      },
    })
  })

  test('object de-structured value', () => {
    const { code, ir } = compileWithVFor(
      '<span v-for={({ id, value }) in items} key={id}>{ id }{ value }</span>',
    )
    expect(code).matchSnapshot()
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'items',
      },
      value: {
        content: '{ id, value }',
        ast: {
          type: 'ObjectExpression',
        },
      },
      key: undefined,
      index: undefined,
    })
  })

  test('object de-structured value (with rest)', () => {
    const { code, ir } = compileWithVFor(
      `<div v-for={(  { id, ...other }, index) in list} key={id}>{ id + other + index }</div>`,
    )
    expect(code).matchSnapshot()
    expect(code).toContain('_getRestElement(_for_item0.value, ["id"])')
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'list',
      },
      value: {
        content: '{ id, ...other }',
        ast: { type: 'ObjectExpression' },
      },
      key: {
        content: 'index',
      },
      index: undefined,
    })
  })

  test('array de-structured value', () => {
    const { code, ir } = compileWithVFor(
      `<div v-for={([id, other], index) in list} key={id}>{ id + other + index }</div>`,
    )
    expect(code).matchSnapshot()
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'list',
      },
      value: {
        content: '[id, other]',
        ast: { type: 'ArrayExpression' },
      },
      key: {
        content: 'index',
      },
      index: undefined,
    })
  })

  test('array de-structured value (with rest)', () => {
    const { code, ir } = compileWithVFor(
      `<div v-for={([id, ...other], index) in list} key={id}>{ id + other + index }</div>`,
    )
    expect(code).matchSnapshot()
    expect(code).toContain('_for_item0.value.slice(1)')
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'list',
      },
      value: {
        content: '[id, ...other]',
        ast: {
          type: 'ArrayExpression',
        },
      },
      key: {
        content: 'index',
      },
      index: undefined,
    })
  })

  test('v-for aliases w/ complex expressions', () => {
    const { code, ir } = compileWithVFor(
      `<div v-for={({ foo, baz: [qux] }) in list}>
        { foo + baz + qux }
      </div>`,
    )
    expect(code).matchSnapshot()
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.FOR,
      source: {
        content: 'list',
      },
      value: {
        content: '{ foo, baz: [qux] }',
        ast: { type: 'ObjectExpression' },
      },
      key: undefined,
      index: undefined,
    })
  })
  test('fast-remove flag', () => {
    const { code, ir } = compileWithVFor(
      `<div>
        <span v-for={j in i}>{ j+i }</span>
      </div>`,
    )

    const op = ir.block.dynamic.children[0].children[1].operation
    expect(op).toMatchObject({
      onlyChild: true,
    })
    expect(code).toMatchSnapshot()
  })

  test('v-for on component', () => {
    const { code, ir } = compileWithVFor(
      `<Comp v-for={item in list}>{item}</Comp>`,
    )
    expect(code).matchSnapshot()
    expect(
      (ir.block.dynamic.children[0].operation as ForIRNode).component,
    ).toBe(true)
  })

  test('v-for on template with single component child', () => {
    const { code, ir } = compileWithVFor(
      `<template v-for={item in list}><Comp>{item}</Comp></template>`,
    )
    expect(code).matchSnapshot()
    expect(
      (ir.block.dynamic.children[0].operation as ForIRNode).component,
    ).toBe(true)
  })
})
