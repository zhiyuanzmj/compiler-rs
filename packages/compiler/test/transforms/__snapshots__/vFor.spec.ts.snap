// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-for > array de-structured value (with rest) 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0, _for_key0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value[0] + _for_item0.value.slice(1) + _for_key0.value))
    return n2
  }, ([id, ...other], index) => (id))
  return n0
"
`;

exports[`compiler: v-for > array de-structured value 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0, _for_key0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value[0] + _for_item0.value[1] + _for_key0.value))
    return n2
  }, ([id, other], index) => (id))
  return n0
"
`;

exports[`compiler: v-for > basic v-for 1`] = `
"
  const n0 = _createFor(() => (items), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value))
    n2.$evtclick = () => remove(_for_item0.value)
    return n2
  }, (item) => (item.id))
  return n0
"
`;

exports[`compiler: v-for > fast-remove flag 1`] = `
"
  const n3 = t1()
  _setInsertionState(n3)
  const n0 = _createFor(() => (i), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value+i))
    return n2
  }, undefined, 1)
  return n3
"
`;

exports[`compiler: v-for > key only binding pattern 1`] = `
"
  const n0 = _createFor(() => (rows), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value.id + _for_item0.value.id))
    return n2
  }, (row) => (row.id))
  return n0
"
`;

exports[`compiler: v-for > multi effect 1`] = `
"
  const n0 = _createFor(() => (items), (_for_item0, _for_key0) => {
    const n2 = t0()
    _renderEffect(() => {
      _setProp(n2, "item", _for_item0.value)
      _setProp(n2, "index", _for_key0.value)
    })
    return n2
  })
  return n0
"
`;

exports[`compiler: v-for > nested v-for 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0) => {
    const n5 = t1()
    _setInsertionState(n5)
    const n2 = _createFor(() => (_for_item0.value), (_for_item1) => {
      const n4 = t0()
      const x4 = _child(n4)
      _setNodes(x4, () => (_for_item1.value+_for_item0.value))
      return n4
    }, undefined, 1)
    return n5
  })
  return n0
"
`;

exports[`compiler: v-for > object de-structured value (with rest) 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0, _for_key0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value.id + _getRestElement(_for_item0.value, ["id"]) + _for_key0.value))
    return n2
  }, ({ id, ...other }, index) => (id))
  return n0
"
`;

exports[`compiler: v-for > object de-structured value 1`] = `
"
  const n0 = _createFor(() => (items), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value.id), () => (_for_item0.value.value))
    return n2
  }, ({ id, value }) => (id))
  return n0
"
`;

exports[`compiler: v-for > object value, key and index 1`] = `
"
  const n0 = _createFor(() => (items), (_for_item0, _for_key0, _for_index0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (id), () => (_for_item0.value), () => (_for_index0.value))
    return n2
  }, (value, key, index) => (id))
  return n0
"
`;

exports[`compiler: v-for > selector pattern 1`] = `
"
  let _selector0_0
  const n0 = _createFor(() => (rows), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _selector0_0(() => {
      _setText(x2, _toDisplayString(selected === _for_item0.value.id ? 'danger' : ''))
    })
    return n2
  }, (row) => (row.id), undefined, ({ createSelector }) => {
    _selector0_0 = createSelector(() => selected)
  })
  return n0
"
`;

exports[`compiler: v-for > selector pattern 2`] = `
"
  let _selector0_0
  const n0 = _createFor(() => (rows), (_for_item0) => {
    const n2 = t0()
    _selector0_0(() => {
      _setClass(n2, selected === _for_item0.value.id ? 'danger' : '')
    })
    return n2
  }, (row) => (row.id), undefined, ({ createSelector }) => {
    _selector0_0 = createSelector(() => selected)
  })
  return n0
"
`;

exports[`compiler: v-for > selector pattern 3`] = `
"
  const n0 = _createFor(() => (rows), (_for_item0) => {
    const n2 = t0()
    _renderEffect(() => _setClass(n2, _for_item0.value.label === _for_item0.value.id ? 'danger' : ''))
    return n2
  }, (row) => (row.id))
  return n0
"
`;

exports[`compiler: v-for > selector pattern 4`] = `
"
  let _selector0_0
  const n0 = _createFor(() => (rows), (_for_item0) => {
    const n2 = t0()
    _selector0_0(() => {
      _setClass(n2, { danger: _for_item0.value.id === selected })
    })
    return n2
  }, (row) => (row.id), undefined, ({ createSelector }) => {
    _selector0_0 = createSelector(() => selected)
  })
  return n0
"
`;

exports[`compiler: v-for > v-for aliases w/ complex expressions 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0) => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (_for_item0.value.foo + baz + _for_item0.value.baz[0]))
    return n2
  })
  return n0
"
`;

exports[`compiler: v-for > v-for on component 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0) => {
    const n3 = _createComponent(Comp)
    const n2 = _child(n3)
    _setNodes(n2, () => (_for_item0.value))
    return [n2, n3]
  }, undefined, 2)
  return n0
"
`;

exports[`compiler: v-for > v-for on template with single component child 1`] = `
"
  const n0 = _createFor(() => (list), (_for_item0) => {
    const n3 = _createComponent(Comp)
    const n2 = _child(n3)
    _setNodes(n2, () => (_for_item0.value))
    return [n2, n3]
  }, undefined, 2)
  return n0
"
`;
